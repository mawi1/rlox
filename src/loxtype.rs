use std::{
    cell::RefCell,
    collections::HashMap,
    fmt::{Debug, Display},
    rc::Rc,
};

use crate::{
    ast::{ClassStatement, FunctionStatement, Statement},
    error::{Error, ErrorDetail},
    interpreter::{run_block, Context, StatementResult},
    Result,
};

pub trait LoxCallable: Debug + Display {
    fn arity(&self) -> usize;
    fn call(&self, arguments: Vec<LoxType>) -> Result<LoxType>;
}

#[derive(Debug)]
pub struct LoxFunction {
    name: String,
    parameters: Vec<String>,
    statements: Rc<Vec<Box<dyn Statement>>>,
    is_initializer: bool,
    ctx: Context,
}

impl LoxFunction {
    pub fn from_statement(
        stmt: &FunctionStatement,
        ctx: Context,
        bind_this: Option<LoxType>,
    ) -> Self {
        let is_initializer = bind_this.is_some() && stmt.name == "init";
        let fn_ctx = if let Some(object) = bind_this {
            let child_ctx = ctx.new_child_ctx();
            child_ctx.define("this", object);
            child_ctx
        } else {
            ctx
        };

        Self {
            name: stmt.name.clone(),
            parameters: stmt.parameters.iter().map(|p| p.name.clone()).collect(),
            statements: stmt.statements.clone(),
            is_initializer,
            ctx: fn_ctx,
        }
    }
}

impl Display for LoxFunction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "<fn \"{}\">", self.name)
    }
}

impl LoxCallable for LoxFunction {
    fn arity(&self) -> usize {
        self.parameters.len()
    }

    fn call(&self, arguments: Vec<LoxType>) -> Result<LoxType> {
        let block_res = run_block(
            self.ctx.clone(),
            &self.statements,
            Some((&self.parameters, arguments)),
        )?;
        if self.is_initializer {
            Ok(self.ctx.get_at(Some(0), "this").unwrap())
        } else {
            match block_res {
                StatementResult::Void => Ok(LoxType::Nil),
                StatementResult::Return(r) => Ok(r),
            }
        }
    }
}

#[derive(Debug)]
pub struct LoxInstance {
    class: Rc<LoxClass>,
    fields: HashMap<String, LoxType>,
}

impl LoxInstance {
    fn new(class: Rc<LoxClass>) -> LoxType {
        LoxType::Instance(Rc::new(RefCell::new(Self {
            class: class.clone(),
            fields: HashMap::new(),
        })))
    }

    pub fn get(instance: Rc<RefCell<LoxInstance>>, name: &str, line: u32) -> Result<LoxType> {
        if let Some(field) = instance.borrow().fields.get(name) {
            return Ok(field.clone());
        }

        instance
            .borrow()
            .class
            .get_method(name, LoxType::Instance(instance.clone()), line)
            .map(|m| LoxType::Callable(Rc::new(m)))
    }

    pub fn set(instance: Rc<RefCell<LoxInstance>>, name: &str, value: LoxType) -> LoxType {
        instance
            .borrow_mut()
            .fields
            .insert(name.to_owned(), value.clone());
        value
    }
}

impl Display for LoxInstance {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} instance", self.class.name)
    }
}

#[derive(Debug)]
pub struct LoxClass {
    pub name: String,
    maybe_superclass: Option<Rc<LoxClass>>,
    methods: Rc<HashMap<String, FunctionStatement>>,
    ctx: Context,
}

impl LoxClass {
    pub fn new(
        stmt: &ClassStatement,
        maybe_superclass: Option<Rc<LoxClass>>,
        ctx: Context,
    ) -> Self {
        let class_ctx = if let Some(superclass) = &maybe_superclass {
            let child_ctx = ctx.new_child_ctx();
            child_ctx.define("super", LoxType::Class(superclass.clone()));
            child_ctx
        } else {
            ctx
        };

        Self {
            name: stmt.name.clone(),
            maybe_superclass,
            methods: stmt.methods.clone(),
            ctx: class_ctx,
        }
    }

    pub fn instantiate(self: Rc<Self>, init_arguments: Vec<LoxType>, line: u32) -> Result<LoxType> {
        let instance = LoxInstance::new(self.clone());

        let maybe_init_method = self.get_method("init", instance.clone(), line).ok();

        let arity = maybe_init_method.as_ref().map_or(0, |i| i.arity());
        if arity != init_arguments.len() {
            return Err(Error::RuntimeError(ErrorDetail::new(
                line,
                format!(
                    "Expected {} arguments but got {}.",
                    arity,
                    init_arguments.len()
                ),
            )));
        }

        if let Some(init_method) = maybe_init_method {
            let _ = init_method.call(init_arguments)?;
        }
        Ok(instance)
    }

    pub fn get_method(&self, name: &str, this: LoxType, line: u32) -> Result<LoxFunction> {
        if let Some(f) = self.methods.get(name) {
            Ok(LoxFunction::from_statement(f, self.ctx.clone(), Some(this)))
        } else {
            self.maybe_superclass.as_ref().map_or_else(
                || {
                    Err(Error::RuntimeError(ErrorDetail::new(
                        line,
                        format!("Undefined property '{}'.", name),
                    )))
                },
                |sc| sc.get_method(name, this, line),
            )
        }
    }
}

impl Display for LoxClass {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

#[derive(Debug, Clone)]
pub enum LoxType {
    Number(f64),
    Boolean(bool),
    String(String),
    Callable(Rc<dyn LoxCallable>),
    Class(Rc<LoxClass>),
    Instance(Rc<RefCell<LoxInstance>>),
    Nil,
}

impl LoxType {
    pub fn is_truthy(&self) -> bool {
        match self {
            LoxType::Number(_) => true,
            LoxType::Boolean(b) => *b,
            LoxType::String(_) => true,
            LoxType::Nil => false,
            LoxType::Callable(_) => true,
            LoxType::Class(_) => true,
            LoxType::Instance(_) => true,
        }
    }
}

impl PartialEq for LoxType {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (LoxType::Number(l), LoxType::Number(r)) => l == r,
            (LoxType::String(l), LoxType::String(r)) => l == r,
            (LoxType::Boolean(l), LoxType::Boolean(r)) => l == r,
            (LoxType::Nil, LoxType::Nil) => true,
            (LoxType::Callable(l), LoxType::Callable(r)) => Rc::ptr_eq(l, r),
            (LoxType::Class(l), LoxType::Class(r)) => Rc::ptr_eq(l, r),
            _ => false,
        }
    }
}

impl Display for LoxType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LoxType::Number(n) => write!(f, "{n}"),
            LoxType::Boolean(b) => write!(f, "{b}"),
            LoxType::String(s) => write!(f, "{s}"),
            LoxType::Nil => write!(f, "nil"),
            LoxType::Callable(c) => {
                write!(f, "{c}")
            }
            LoxType::Class(c) => write!(f, "{c}"),
            LoxType::Instance(i) => write!(f, "{}", i.borrow()),
        }
    }
}
