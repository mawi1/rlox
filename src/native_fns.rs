use std::{
    fmt::Display,
    time::{SystemTime, UNIX_EPOCH},
};

use crate::{LoxCallable, LoxType};

#[derive(Debug)]
pub struct Clock();

impl Display for Clock {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "<native fn clock>")
    }
}

impl LoxCallable for Clock {
    fn arity(&self) -> usize {
        0
    }

    fn call(&self, _arguments: Vec<LoxType>) -> crate::Result<LoxType> {
        let now = SystemTime::now();
        let elapsed = now.duration_since(UNIX_EPOCH).unwrap();
        Ok(LoxType::Number(elapsed.as_secs() as f64))
    }
}
