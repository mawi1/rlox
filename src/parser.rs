use std::{iter::Peekable, rc::Rc};

use crate::{
    ast::*,
    error::{Error, ErrorDetail},
    loxtype::LoxType,
    token::{
        Literal, Token,
        TokenType::{self, *},
    },
    Result,
};

#[derive(Debug)]
enum FunctionKind {
    Function,
    #[allow(dead_code)]
    Method,
}

pub struct Parser<'a> {
    tokens: Peekable<std::iter::Take<std::slice::Iter<'a, Token>>>,
    errors: Vec<ErrorDetail>,
    last_line: u32,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: &'a Vec<Token>) -> Self {
        Self {
            // iterate without Eof token at end
            tokens: tokens.iter().take(tokens.len() - 1).peekable(),
            errors: Vec::new(),
            last_line: tokens
                .get(tokens.len().wrapping_sub(2))
                .map(|t| t.line)
                .unwrap_or(1),
        }
    }

    pub fn parse(mut self) -> Result<Vec<Box<dyn Statement>>> {
        let mut statements = vec![];

        while self.tokens.peek().is_some() {
            match self.declaration() {
                Ok(s) => statements.push(s),
                Err(e) => {
                    self.errors.push(e);
                    self.synchronize()
                }
            }
        }

        if self.errors.is_empty() {
            Ok(statements)
        } else {
            Err(Error::SyntaxErrors(self.errors))
        }
    }

    fn synchronize(&mut self) {
        while let Some(token) = self.tokens.peek() {
            let ty = token.ty;
            if ty == Semicolon {
                self.tokens.next();
                return;
            }
            if [Class, Fun, Var, For, If, While, Print, Return]
                .iter()
                .any(|&tt| tt == ty)
            {
                return;
            };
            self.tokens.next();
        }
    }

    fn match_token_type(&mut self, tt: TokenType) -> Option<&'a Token> {
        self.tokens.next_if(|t| t.ty == tt)
    }

    fn match_token_types(&mut self, tts: &[TokenType]) -> Option<&'a Token> {
        if let Some(t) = self.tokens.peek() {
            if tts.iter().any(|&tt| tt == t.ty) {
                Some(self.tokens.next().unwrap())
            } else {
                None
            }
        } else {
            None
        }
    }

    fn is_next_token_type(&mut self, tt: TokenType) -> bool {
        self.tokens.next_if(|t| t.ty == tt).is_some()
    }

    fn consume(&mut self, token_ty: TokenType) -> std::result::Result<&'a Token, ErrorDetail> {
        if let Some(n) = self.tokens.peek() {
            if n.ty == token_ty {
                Ok(self.tokens.next().unwrap())
            } else {
                Err(ErrorDetail::new(n.line, format!("Expect '{token_ty}'.")))
            }
        } else {
            Err(ErrorDetail::new(
                self.last_line,
                format!("Expect '{token_ty}'."),
            ))
        }
    }

    fn declaration(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        match self.tokens.peek().unwrap().ty {
            Var => self.var_declaration(),
            Fun => self.function(FunctionKind::Function),
            _ => self.statement(),
        }
    }

    fn function(
        &mut self,
        kind: FunctionKind,
    ) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        let fun_token = self.tokens.next().unwrap();
        if let Some(name_token) = self.tokens.next_if(|t| t.ty == Identifier) {
            let name = name_token.lexeme.clone();

            self.consume(LeftParen)?;
            let mut parameters = vec![];
            if self.tokens.peek().is_some_and(|t| t.ty != RightParen) {
                loop {
                    let identifier = self.consume(Identifier)?;
                    parameters.push(Parameter {
                        name: identifier.lexeme.clone(),
                        line: identifier.line,
                    });
                    if !self.is_next_token_type(Comma) {
                        break;
                    }
                }
            }
            let paren_token = self.consume(RightParen)?;
            if parameters.len() > 255 {
                self.errors.push(ErrorDetail::new(
                    paren_token.line,
                    "Can't have more than 255 parameters.",
                ));
            }

            self.consume(LeftBrace)?;
            let block = self.block_statement()?;

            Ok(Box::new(FunctionStatement {
                name,
                parameters,
                statements: Rc::new(block.statements),
                line: fun_token.line,
            }))
        } else {
            let message = match kind {
                FunctionKind::Function => "Expect function name.",
                FunctionKind::Method => "Expect method name.",
            };
            Err(ErrorDetail::new(fun_token.line, message))
        }
    }

    fn var_declaration(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        let var_token = self.tokens.next().unwrap();
        let name = self.consume(Identifier)?;

        let initializer = if self.is_next_token_type(Equal) {
            Some(self.expression()?)
        } else {
            None
        };
        self.consume(Semicolon)?;

        Ok(Box::new(VarStatement {
            name: name.lexeme.clone(),
            initializer: initializer,
            line: var_token.line,
        }))
    }

    fn statement(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        match self.tokens.peek().unwrap().ty {
            For => {
                self.tokens.next();
                self.for_statement()
            }
            If => {
                self.tokens.next();
                self.if_statement()
            }
            LeftBrace => {
                self.tokens.next();
                self.block_statement()
                    .map(|b| Box::new(b) as Box<dyn Statement>)
            }
            Print => self.print_statement(),
            Return => self.return_statemen(),
            While => {
                self.tokens.next();
                self.while_statement()
            }
            _ => self.expression_statement(),
        }
    }

    fn return_statemen(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        let return_token = self.tokens.next().unwrap();
        let maybe_expression = match self.tokens.peek().is_some_and(|t| t.ty != Semicolon) {
            true => Some(self.expression()?),
            false => None,
        };
        self.consume(Semicolon)?;
        Ok(Box::new(ReturnStatement {
            maybe_expression,
            line: return_token.line,
        }))
    }

    fn for_statement(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        self.consume(LeftParen)?;

        let opt_initializer = if self.is_next_token_type(Semicolon) {
            None
        } else if self.tokens.peek().is_some_and(|t| t.ty == Var) {
            Some(self.var_declaration()?)
        } else {
            Some(self.expression_statement()?)
        };

        let opt_for_condition = if self.is_next_token_type(Semicolon) {
            None
        } else {
            let e = self.expression()?;
            self.consume(Semicolon)?;
            Some(e)
        };

        let opt_increment = if self.is_next_token_type(RightParen) {
            None
        } else {
            let i = self.expression()?;
            self.consume(RightParen)?;
            Some(i)
        };

        let for_body = self.statement()?;

        //desugar as while-loop:
        //{
        // initializer;
        // while(condition) {
        //  body;
        //  increment;
        // }
        //}
        let condition =
            opt_for_condition.unwrap_or(Box::new(LiteralExpression(LoxType::Boolean(true))));

        let mut body_statements: Vec<Box<dyn Statement>> = vec![for_body];
        if let Some(increment) = opt_increment {
            body_statements.push(Box::new(ExpressionStatement(increment)));
        }
        let body = Box::new(BlockStatement {
            statements: body_statements,
        });

        let while_statement = Box::new(WhileStatement { condition, body });
        let mut block_statements: Vec<Box<dyn Statement>> = vec![];
        if let Some(initializer) = opt_initializer {
            block_statements.push(initializer);
        }
        block_statements.push(while_statement);

        Ok(Box::new(BlockStatement {
            statements: block_statements,
        }))
    }

    fn while_statement(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        self.consume(LeftParen)?;
        let condition = self.expression()?;
        self.consume(RightParen)?;
        let body = self.statement()?;
        Ok(Box::new(WhileStatement { condition, body }))
    }

    fn if_statement(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        self.consume(LeftParen)?;
        let condition = self.expression()?;
        self.consume(RightParen)?;

        let then_branch = self.statement()?;
        let else_branch = self
            .match_token_type(Else)
            .map(|_| self.statement())
            .transpose()?;
        Ok(Box::new(IfStatement {
            condition,
            then_branch,
            else_branch,
        }))
    }

    fn print_statement(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        let print_token = self.tokens.next().unwrap();
        let expression = self.expression()?;
        self.consume(Semicolon)?;
        Ok(Box::new(PrintStatement {
            expression,
            line: print_token.line,
        }))
    }

    fn block_statement(&mut self) -> std::result::Result<BlockStatement, ErrorDetail> {
        let mut statements = Vec::new();

        while let Some(token) = self.tokens.peek() {
            if token.ty == RightBrace {
                break;
            } else {
                statements.push(self.declaration()?);
            }
        }

        self.consume(RightBrace)?;
        Ok(BlockStatement { statements })
    }

    fn expression_statement(&mut self) -> std::result::Result<Box<dyn Statement>, ErrorDetail> {
        let e = self.expression()?;
        self.consume(Semicolon)?;
        Ok(Box::new(ExpressionStatement(e)))
    }

    fn expression(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        self.assignment()
    }

    fn assignment(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let expr = self.or()?;

        if let Some(eq_token) = self.match_token_type(Equal) {
            let value = self.assignment()?;

            let expr_any = expr.as_any();
            match expr_any.downcast_ref::<VariableExpression>() {
                Some(var_expr) => {
                    return Ok(Box::new(AssignExpression {
                        name: var_expr.name.clone(),
                        value: value,
                        maybe_distance: None,
                        line: eq_token.line,
                    }));
                }
                None => {
                    self.errors.push(ErrorDetail::new(
                        eq_token.line,
                        "Invalid assignment target.",
                    ));
                }
            }
        }
        Ok(expr)
    }

    fn or(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.and()?;

        while self.is_next_token_type(Or) {
            let right = self.and()?;
            expr = Box::new(LogicalExpression {
                left: expr,
                right: right,
                operator: LogicalOperator::Or,
            });
        }
        return Ok(expr);
    }

    fn and(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.equality()?;

        while self.is_next_token_type(And) {
            let right = self.equality()?;
            expr = Box::new(LogicalExpression {
                left: expr,
                right: right,
                operator: LogicalOperator::And,
            });
        }
        return Ok(expr);
    }

    fn equality(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.comparison()?;

        while let Some(operator) = self.match_token_types(&[BangEqual, EqualEqual]) {
            let right = self.comparison()?;
            expr = match operator.ty {
                BangEqual => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::NotEqual,
                    line: operator.line,
                }),
                EqualEqual => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Equal,
                    line: operator.line,
                }),
                _ => unreachable!(),
            };
        }
        Ok(expr)
    }

    fn comparison(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.term()?;

        while let Some(operator) = self.match_token_types(&[Greater, GreaterEqual, Less, LessEqual])
        {
            let right = self.term()?;
            expr = match operator.ty {
                Greater => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Greater,
                    line: operator.line,
                }),
                GreaterEqual => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::GreaterOrEqual,
                    line: operator.line,
                }),
                Less => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Less,
                    line: operator.line,
                }),
                LessEqual => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::LessOrEqual,
                    line: operator.line,
                }),
                _ => unreachable!(),
            };
        }

        Ok(expr)
    }

    fn term(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.factor()?;

        while let Some(operator) = self.match_token_types(&[Minus, Plus]) {
            let right = self.factor()?;
            expr = match operator.ty {
                Minus => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Substract,
                    line: operator.line,
                }),
                Plus => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Add,
                    line: operator.line,
                }),
                _ => unreachable!(),
            };
        }

        Ok(expr)
    }

    fn factor(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.unary()?;

        while let Some(operator) = self.match_token_types(&[Star, Slash]) {
            let right = self.unary()?;
            expr = match operator.ty {
                Star => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Multiply,
                    line: operator.line,
                }),
                Slash => Box::new(BinaryExpression {
                    left: expr,
                    right: right,
                    operator: BinaryOperator::Divide,
                    line: operator.line,
                }),
                _ => unreachable!(),
            };
        }

        Ok(expr)
    }

    fn unary(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        if let Some(operator) = self.match_token_types(&[Bang, Minus]) {
            let expression = self.unary()?;

            return Ok(match operator.ty {
                Bang => Box::new(NotExpression(expression)),
                Minus => Box::new(NegExpression {
                    expression,
                    line: operator.line,
                }),
                _ => unreachable!(),
            });
        }

        self.call()
    }

    fn finish_call(
        &mut self,
        callee: Box<dyn Expression>,
    ) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut arguments = vec![];

        if self.tokens.peek().is_some_and(|t| t.ty != RightParen) {
            loop {
                arguments.push(self.expression()?);
                if !self.is_next_token_type(Comma) {
                    break;
                }
            }
        }
        let paren_token = self.consume(RightParen)?;
        if arguments.len() > 255 {
            self.errors.push(ErrorDetail::new(
                paren_token.line,
                "Can't have more than 255 arguments.",
            ));
        }

        Ok(Box::new(CallExpression {
            callee,
            arguments,
            line: paren_token.line,
        }))
    }

    fn call(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        let mut expr = self.primary()?;

        loop {
            if self.is_next_token_type(LeftParen) {
                expr = self.finish_call(expr)?;
            } else {
                break;
            }
        }

        Ok(expr)
    }

    fn primary(&mut self) -> std::result::Result<Box<dyn Expression>, ErrorDetail> {
        if let Some(token) = self.tokens.next() {
            match token.ty {
                Nil => Ok(Box::new(NilExpression())),
                True => Ok(Box::new(LiteralExpression(LoxType::Boolean(true)))),
                False => Ok(Box::new(LiteralExpression(LoxType::Boolean(false)))),
                Number => {
                    if let Literal::Number(n) = token.literal.as_ref().expect("no literal value") {
                        Ok(Box::new(LiteralExpression(LoxType::Number(*n))))
                    } else {
                        panic!("literal type mismatch");
                    }
                }
                String => {
                    if let Literal::String(s) = token.literal.as_ref().expect("no literal value") {
                        Ok(Box::new(LiteralExpression(LoxType::String(s.clone()))))
                    } else {
                        panic!("literal type mismatch");
                    }
                }
                LeftParen => {
                    let expr = self.expression()?;
                    self.consume(RightParen)?;
                    Ok(Box::new(GroupingExpression(expr)))
                }
                Identifier => Ok(Box::new(VariableExpression {
                    name: token.lexeme.clone(),
                    maybe_distance: None,
                    line: token.line,
                })),
                _ => Err(ErrorDetail::new(token.line, "Expect expression.")),
            }
        } else {
            Err(ErrorDetail::new(self.last_line, "Expect expression."))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use insta::{assert_debug_snapshot, glob};

    use crate::scanner::scan_tokens;

    use super::*;

    #[test]
    fn test_parser() {
        glob!("../test_programs/parsing/", "**/*.lox", |path| {
            let input = fs::read_to_string(path).unwrap();
            let tokens = scan_tokens(&input).unwrap();
            let parser = Parser::new(&tokens);
            assert_debug_snapshot!(parser.parse());
        });
    }
}
