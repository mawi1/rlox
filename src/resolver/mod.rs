mod resolve_expr;
mod resolve_stmt;

use std::collections::HashMap;

use crate::ast::Statement;
use crate::error::{Error, ErrorDetail};
use crate::Result;

#[derive(Debug, PartialEq, Eq)]
enum FunctionType {
    Function,
}

#[derive(Debug, PartialEq, Eq)]
enum VariableState {
    Declared,
    Defined,
}

pub(crate) struct Scopes {
    scopes: Vec<HashMap<String, VariableState>>,
    function_types: Vec<FunctionType>,
    errors: Vec<ErrorDetail>,
}

impl Scopes {
    pub fn new() -> Self {
        Self {
            scopes: vec![],
            function_types: vec![],
            errors: vec![],
        }
    }

    pub fn begin_scope(&mut self) {
        self.scopes.push(HashMap::new());
    }

    pub fn end_scope(&mut self) {
        self.scopes.pop();
    }

    pub fn begin_function(&mut self) {
        self.function_types.push(FunctionType::Function);
    }

    pub fn end_function(&mut self) {
        self.function_types.pop();
    }

    pub fn declare(&mut self, name: &str, line: u32) {
        if let Some(hm) = self.scopes.last_mut() {
            if hm.contains_key(name) {
                self.errors.push(ErrorDetail::new(
                    line,
                    "Already a variable with this name in this scope.",
                ));
            } else {
                hm.insert(name.to_owned(), VariableState::Declared);
            }
        }
    }

    pub fn define(&mut self, name: &str) {
        if let Some(hm) = self.scopes.last_mut() {
            hm.insert(name.to_owned(), VariableState::Defined);
        }
    }

    pub fn check_return_statement(&mut self, line: u32) {
        if self.function_types.len() == 0 {
            self.errors
                .push(ErrorDetail::new(line, "Can't return from top-level code."));
        }
    }

    pub fn check_initialized(&mut self, name: &str, line: u32) {
        if self
            .scopes
            .last()
            .is_some_and(|hm| hm.get(name).is_some_and(|v| *v == VariableState::Declared))
        {
            self.errors.push(ErrorDetail::new(
                line,
                "Can't read local variable in its own initializer.",
            ));
        }
    }

    pub fn resolve_local(&self, name: &str) -> Option<u32> {
        self.scopes
            .iter()
            .rev()
            .position(|hm| hm.contains_key(name))
            .map(|v| v as u32)
    }

    pub fn into_errors(self) -> Vec<ErrorDetail> {
        self.errors
    }
}

pub trait Resolve {
    fn resolve(&mut self, scopes: &mut Scopes);
}

pub fn resolve(statements: &mut [Box<dyn Statement>]) -> Result<()> {
    let mut scopes = Scopes::new();
    for statement in statements {
        statement.resolve(&mut scopes)
    }

    let errors = scopes.into_errors();
    if errors.len() > 0 {
        Err(Error::ResolverErrors(errors))
    } else {
        Ok(())
    }
}
