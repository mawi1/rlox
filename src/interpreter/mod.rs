mod env;
mod eval;
mod exec;

use std::cell::RefCell;
use std::io::{stdout, Stdout};
use std::rc::Rc;

use crate::ast::Statement;
use crate::loxtype::LoxType;
use crate::native_fns::Clock;
use crate::parser::Parser;
use crate::resolver::resolve;
use crate::scanner::scan_tokens;
use crate::Result;

pub use self::env::{Environment, UndefinedVariable};

pub enum StatementResult {
    Void,
    Return(LoxType),
}

#[derive(Debug, Clone)]
pub struct Context {
    globals: Rc<RefCell<Environment>>,
    env: Rc<RefCell<Environment>>,
    stout: Rc<RefCell<Stdout>>,
    #[cfg(test)]
    test_stout: Rc<RefCell<String>>,
}

impl Context {
    pub fn new() -> Self {
        let globals = Environment::new(None);
        let env = globals.clone();
        Self {
            globals,
            env,
            stout: Rc::new(RefCell::new(stdout())),
            #[cfg(test)]
            test_stout: Rc::new(RefCell::new(String::new())),
        }
    }

    pub fn define(&self, name: &str, value: LoxType) {
        self.env.borrow_mut().define(name, value);
    }

    pub fn assign_at(
        &self,
        maybe_distance: Option<u32>,
        name: &str,
        value: LoxType,
    ) -> std::result::Result<(), UndefinedVariable> {
        if let Some(distance) = maybe_distance {
            self.env.borrow_mut().assign_at(distance, name, value)
        } else {
            self.globals.borrow_mut().assign_at(0, name, value)
        }
    }

    pub fn get_at(
        &self,
        maybe_distance: Option<u32>,
        name: &str,
    ) -> std::result::Result<LoxType, UndefinedVariable> {
        if let Some(distance) = maybe_distance {
            self.env.borrow().get_at(distance, name)
        } else {
            self.globals.borrow().get_at(0, name)
        }
    }

    #[cfg(not(test))]
    pub fn write_stdout(&self, t: &str) -> std::result::Result<(), std::io::Error> {
        use std::io::Write;

        let mut out = self.stout.borrow_mut();
        out.write_all(t.as_bytes()).and_then(|_| out.flush())
    }

    #[cfg(test)]
    pub fn write_stdout(&self, t: &str) -> std::result::Result<(), std::io::Error> {
        self.test_stout.borrow_mut().push_str(t);
        Ok(())
    }

    pub fn new_child_ctx(&self) -> Self {
        Context {
            globals: self.globals.clone(),
            env: Environment::new(Some(self.env.clone())),
            stout: self.stout.clone(),
            #[cfg(test)]
            test_stout: self.test_stout.clone(),
        }
    }

    #[cfg(test)]
    pub fn into_writer(self) -> String {
        self.test_stout.borrow().clone()
    }
}

pub trait Eval {
    fn eval(&self, ctx: Context) -> Result<LoxType>;
}

pub trait Exec {
    fn exec(&self, ctx: Context) -> Result<StatementResult>;
}

pub(crate) fn run_block(
    ctx: Context,
    statements: &[Box<dyn Statement>],
    maybe_params_args: Option<(&[String], Vec<LoxType>)>,
) -> crate::Result<StatementResult> {
    let block_ctx = ctx.new_child_ctx();
    if let Some((params, args)) = maybe_params_args {
        assert!(params.len() == args.len(), "");
        for (param, arg) in params.into_iter().zip(args) {
            block_ctx.define(param, arg);
        }
    }
    for statement in statements.iter() {
        if let StatementResult::Return(r) = statement.exec(block_ctx.clone())? {
            return Ok(StatementResult::Return(r));
        }
    }
    Ok(StatementResult::Void)
}
pub struct Interpreter {
    ctx: Context,
}

impl Interpreter {
    pub fn new() -> Self {
        let ctx = Context::new();
        ctx.define("clock", LoxType::Callable(Rc::new(Clock())));
        Self { ctx }
    }

    pub fn run(&self, source: &str) -> Result<()> {
        let tokens = scan_tokens(source)?;
        let mut statements = Parser::new(&tokens).parse()?;
        resolve(&mut statements)?;

        for statement in statements {
            statement.exec(self.ctx.clone())?;
        }
        Ok(())
    }

    #[cfg(test)]
    pub fn get_output(self) -> String {
        self.ctx.into_writer()
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use insta::{assert_snapshot, glob};

    use super::*;

    #[test]
    fn test_interpreter() {
        glob!("../../test_programs/interpreter/", "**/*.lox", |path| {
            let input = fs::read_to_string(path).unwrap();
            let interpreter = Interpreter::new();
            let output = match interpreter.run(&input) {
                Ok(_) => interpreter.get_output(),
                Err(e) => e.to_string(),
            };
            assert_snapshot!(output);
        });
    }
}
