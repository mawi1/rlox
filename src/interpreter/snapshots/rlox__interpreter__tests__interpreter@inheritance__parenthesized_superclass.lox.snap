---
source: src/interpreter/mod.rs
expression: output
input_file: test_programs/interpreter/inheritance/parenthesized_superclass.lox
---
Syntax error(s):
[ line 4 ] : Expect 'Identifier'.
